### PR TITLE
Union Access

### PR DESCRIPTION
**Pull request by FintanH**
Addressing phase 1 of https://github.com/dhall-lang/dhall-lang/issues/244 and https://github.com/dhall-lang/dhall-haskell/issues/635 

@Gabriel439: I haven't implemented any tests for this and I should but is there documentation of how the tests are laid out? I'm slightly confused but possibly it'd be fine if I spent some more time looking through them :)

### CODE REVIEW

**Comment by f-f** [comment]
@FintanH there is some info about how tests are structured in https://github.com/dhall-lang/dhall-lang/issues/240. I'm writing some of that down in a file to keep in that repo (because we'll have "assurance tests" there)

**Comment by FintanH** [comment]
@f-f have I told you you're a hero? Thank you. I think I can see what's going on now :)

**Comment by FintanH** [comment]
Is there something I have to do when adding new files? It seems the tests couldn't find the new files, but I added them :/

**Comment by f-f** [comment]
@FintanH yes, you should add `tests/normalization/tutorial/access/*` somewhere around [here](https://github.com/dhall-lang/dhall-haskell/blob/master/dhall.cabal#L31)

File: src/Dhall/TypeCheck.hs [hunk 1]
```diff
@@ -3273,22 +3293,70 @@ prettyTypeMessage (NotARecord lazyText0 expr0 expr1) = ErrorMessages {..}
         \      Invalid: Not a record                                                     \n\
         \                                                                                \n\
         \                                                                                \n\
+        \────────────────────────────────────────────────────────────────────────────────\n\
+        \                                                                                \n\
+        \You tried to access the field:aaa                                               \n\
```
Comment (Gabriel439) [inline src/Dhall/TypeCheck.hs]:
```suggestion
        \You tried to access the field:                                                  \n\
```

Comment (FintanH) [reply src/Dhall/TypeCheck.hs]: Hahaha whooops!

**Comment by FintanH** [comment]
@f-f Ya I thought about that, but didn't see any existing ones that had similar functionality. But that's a bad excused really :flushed:

File: src/Dhall/TypeCheck.hs [hunk 2]
```diff
@@ -3273,22 +3293,70 @@ prettyTypeMessage (NotARecord lazyText0 expr0 expr1) = ErrorMessages {..}
...
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌─────────────────────────────────┐                                         \n\
+        \    │ { foo = True, bar = \"ABC\", baz = 1 }.{ foo, bar } │  This is valid ...  \n\
+        \    └─────────────────────────────────┘                                         \n\
+        \                                                                                \n\
+        \                                                                                \n\
+        \    ┌───────────────────────────────────────────┐                               \n\
+        \    │ λ(r : { foo : Bool, bar : Text , baz : Natural }) → r.{ foo, bar } │  ... and so is this           \n\
+        \    └───────────────────────────────────────────┘                               \n\
```
Comment (ocharles) [inline src/Dhall/TypeCheck.hs]: Is this boxing correct? It doesn't render correctly, but that could be my browser (I use a font with ligatures that might be messing things up, though everything else looks correct.

**Pull request merged**